### ISSUE TITLE
Function empty_aligned fails on certain dtypes...

### ISSUE THREAD

**Issue by TrekkieByDay** [pyFFTW/pyFFTW#111]
When building the FFT, the function `builders._utils._Xfftn` is eventually called. This calls `builders._utils._compute_array_shapes` to determine the appropriate `input_shape` and `output_shape` and then eventually `pyfftw.empty_aligned(output_shape, output_dtype)`. We were getting the following exception when building FFTs on very large `complex64` arrays:

```
ValueError: negative dimensions are not allowed
```

This was eventually traced to the `builders._compute_array_shapes` call which produced a tuple of `int32` dimensions for the respective shape tuples. I assume they were `int32` as a consequence of working on `complex64` arrays. In any event, the negative dimensions were probably due to integer wrapping for numbers greater than the capacity of `int32`. See the following instructive example, which was carried out using the pre-built `pyFFTW-0.10.1-cp34-none-win_amd64.whl`::

```
>>> dtype = np.dtype(np.complex64)
>>> shape = (np.int32(2**14), np.int32(2**14))
>>> pyfftw.empty_aligned(shape, dtype)

ValueError: negative dimensions are not allowed
```

And a little bit of tweaking...

```
>>> shape = (np.int32(2**14 - 1), np.int32(2**14))
>>> pyfftw.empty_aligned(shape, dtype)
array([[ 0.+0.j,  0.+0.j,  0.+0.j, ...,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j, ...,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j, ...,  0.+0.j,  0.+0.j,  0.+0.j],
       ..., 
       [ 0.+0.j,  0.+0.j,  0.+0.j, ...,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j, ...,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j, ...,  0.+0.j,  0.+0.j,  0.+0.j]], dtype=complex64)
```

In order to perform the required FFTs, I changed `builders._compute_array_shapes` so that it **always** casts the shape to a tuple of Python `int`s rather than something based on the provided `dtype`. For us anyway, there does not seem to be a good reason for the shape tuple to be anything other than high capacity integers... the memory footprint is obviously dominated by the `dtype` of the actual array rather than that of its metadata.

I just wanted to bring this to your attention... I didn't know if this was expected behavior or not. In any event, we did have to make a hack in your package, which we always try to avoid! :)

**Comment by hgomersall** [comment]
Uhm, your example doesn't work for me (the array is created just ine in both cases) - I assume you mean `2**31` as the shape arguments?

That said, I have no idea why the shape should be 32-bit - AFAICT it's never coerced to be such, and in any case, that would not be right. What platform etc are you running?

How big actually is your array? Issue #14 needs to be fixed at some point (I think it's quite easy to do, but I've never had the impetus), and that is surely the more pertinent issue in your case? As currently implemented, an array cannot be [bigger than sizes represented by an int](http://www.fftw.org/fftw3_doc/Guru-vector-and-transform-sizes.html#Guru-vector-and-transform-sizes).

**Comment by TrekkieByDay** [comment]
My platforrm is 64 bit Windows, Python 3.4. I linked the build I am using. In the compute_array_shapes function, the elements of the shape tuple are indeed cast to Numpy int32s. Also, the examples I provided are exact... you should be able to copy paste them. I have produced this response on multiple computers (all with the same build as mentioned above).

**Comment by hgomersall** [comment]
Ah, yes apologies for not noticing your build string.

Arrays of size `2**14` should work just fine, and do so on my machine - you shouldn't have a problem until you get `2**17` times bigger than that.

I understand your types might be `int32`, but AFAICT it's not being set anywhere in `pyfftw`. There is nothing in `builders` that does that cast.

This really doesn't make sense to me. Even if the type is `int32` that doesn't explain it. Even `int16` should be fine as I understand how you've described it. If `shape = (np.int32(2**14 - 1), np.int32(2**14))` is producing negative values, that is a bug with your numpy installation.

**Comment by hgomersall** [comment]
What should happen:
```
In [1]: (np.int32(2**14), np.int32(2**14))
Out[1]: (16384, 16384)

In [2]: (np.int32(2**31-1), np.int32(2**31-1))
Out[2]: (2147483647, 2147483647)

In [3]: (np.int32(2**31), np.int32(2**31))
Out[3]: (-2147483648, -2147483648)
```

**Comment by TrekkieByDay** [comment]
Note that you are not purposefully casting it to `int32`. That is just the type that Numpy returns when you query the shape. In the function `_compute_array_shapes`, you return the following:

    return tuple(input_shape), tuple(output_shape)

The elements of `input_shape` and `output_shape` in my case are `int32`. To fix the problem, I do something like this:

    input_shape = tuple([int(num) for num in input_shape])
    output_shape = tuple([int(num) for num in output_shape])
    return input_shape, output_shape

Note that I have reproduced this error on both my home and work computers (though they are all 64 bit Windows using the same PyFFTW build). I'm sorry that you haven't yet been able to reproduce this on your end. I'm not crazy though... it really is happening. :) Also, I don't believe it's being caused by a bug in Numpy. Numpy is behaving just as I expect it to.  In any event, thanks for looking into this.

**Comment by hgomersall** [comment]
Ah, I think I understand what the problem is. Are you able at all to create an `int64`?

Consider the following code:
https://github.com/pyFFTW/pyFFTW/blob/master/pyfftw/utils.pxi#L192

If you run that little loop, what happens? The issue seems to be that your default int type in python is an int32, which it really shouldn't be, but windows has all manor of these foibles. It's probably related to `http://stackoverflow.com/questions/36278590/numpy-array-dtype-is-coming-as-int32-by-default-in-a-windows-10-64-bit-machine`.

**Comment by hgomersall** [comment]
[This](http://stackoverflow.com/questions/30357552/is-there-any-way-to-force-python-to-use-64-bit-integers-on-windows) seems to be the relevant reference to normal python ints.

Can you create `np.int64` types ok? The simple fix would be to make `array_length` on line 193 a `np.int64`.

**Comment by hgomersall** [comment]
FYI, I would have seen it a bit quicker if you'd shown me what `shape` contains after `shape = (np.int32(2**14), np.int32(2**14))`. The fact that its 32-bit is a red herring (albeit due to the same problem).

**Comment by TrekkieByDay** [comment]
Wow... good to know. Call me nuts, but I just assumed that Python integers would default to 64 bit on a 64 bit machine! Yes I can make `np.int64` no problem. Ahh, sorry. It's always tricky to know exactly what to show in these cases. But yes, I think explicitly forcing `int64` would be the solution.

**Comment by hgomersall** [comment]
Can you try running that loop to check the `int64` solution works. It would save me having to reboot :)

**Comment by TrekkieByDay** [comment]
What exactly do you want me to run?


    # Apparently there is an issue with numpy.prod wrapping around on 32-bits
    # on Windows 64-bit. This shouldn't happen, but the following code
    # alleviates the problem.
    if not isinstance(shape, (int, np.integer)):
        array_length = 1
        for each_dimension in shape:
            array_length *= each_dimension

**Comment by hgomersall** [comment]
```
array_length = np.int64(1)
for each_dimension in shape:
    array_length *= each_dimension

print(array_length)
```

Also, it would be good to try it without the `np.int64` as a sanity check.

**Comment by hgomersall** [comment]
Also, can you try:
```
print(np.prod(np.int64(shape)))
```

**Comment by TrekkieByDay** [comment]
import numpy as np
    
     x = np.arange(10, dtype='complex64')
     shape = x.shape
     array_length = np.int64(1)
     for each_dimension in shape:
          array_length *= each_dimension

     print(array_length)
     
     print(np.prod(np.int64(shape)))

     >>>>
     10
     10

**Comment by hgomersall** [comment]
I mean on the sizes you're using!

**Comment by TrekkieByDay** [comment]
import numpy as np

    x = np.arange(2**28, dtype='complex64').reshape((2**14, 2**14))
    shape = x.shape
    array_length = np.int64(1)
    for each_dimension in shape:
        array_length *= each_dimension
    
    print(array_length)
    
    print(np.prod(np.int64(shape)))
    
    >>>>>
    268435456
    268435456

**Comment by hgomersall** [comment]
Actually I think the failing might be on line 201.

So, can you run the following:

```python
x = np.zeros((2**14, 2**14), dtype='complex64')
itemsize = x.itemsize

shape = x.shape
int64_array_length = np.int64(1)
orig_array_length = 1
n = 32

for each_dimension in shape:
    int64_array_length *= each_dimension
    orig_array_length *= each_dimension

print('original: ' + str(orig_array_length))
print('original new array bytes: ' + str(orig_array_length * itemsize + n))
print('int64 array: ' + str(int64_array_length))
print('int64 new array bytes: ' + str(int64_array_length * itemsize + n))
print('int64 new array bytes modified: ' + str(int64_array_length * np.int64(itemsize) + np.int64(n)))
print('int64 new array bytes alt: ' + str(np.prod(np.int64(shape)) * itemsize + n))
```

**Comment by TrekkieByDay** [comment]
Sorry was getting ready for work. Here you go:

    >>>>
    original: 268435456
    original new array bytes: 2147483680
    int64 array: 268435456
    int64 new array bytes: 2147483680
    int64 new array bytes modified: 2147483680
    int64 new array bytes alt: 2147483680

**Comment by hgomersall** [comment]
Well that's intriguing - all those sizes are as expected (and bigger than ``2**31``). I wonder if it's a cython thing - I'll need to have a play in windows and that really upsets me.

**Comment by TrekkieByDay** [comment]
Lol you're like my son... dirty Windows! But yeah I feel like after talking to you that this all traces back to some Windows quirk. :( In any event, thank you for checking into this... and just to be clear, the fix I mentioned above is a suitable bandaid... we aren't losing compute time at the moment, so that's good. :dancer:

**Comment by hyliu1989** [comment]
I recently encounter this problem. I guess a good fix can be replacing
`array_length = 1`
with a typed statement (some `cdef`)? I am not familiar with cython language so I am not sure what statement to use exactly.

**Comment by grlee77** [comment]
closed by #192
